data race in priority and fairness unit test

### What happened?

`TestApfWatchHandlePanic/post-execute_panic` has a data race

```
{Failed  === RUN   TestApfWatchHandlePanic/post-execute_panic
==================
WARNING: DATA RACE
Read at 0x000003a22108 by goroutine 426:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:179 +0x584
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x547
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.TestApfWatchHandlePanic.func3.1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:598 +0xd0
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /usr/local/go/src/net/http/server.go:3071 +0x58

Previous write at 0x000003a22108 by goroutine 166:
  sync/atomic.AddInt32()
      /usr/local/go/src/runtime/race_amd64.s:279 +0xb
  sync/atomic.AddInt32()
      <autogenerated>:1 +0x1a
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1.7.2()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:204 +0x3e
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:436 +0x32
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*fakeWatchApfFilter).Handle()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:415 +0xa6
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1.8()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:260 +0x1a4

Goroutine 426 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3071 +0x80c
  net/http/httptest.(*Server).goServe.func1()
      /usr/local/go/src/net/http/httptest/server.go:308 +0xb2

Goroutine 166 (finished) created at:
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithPriorityAndFairness.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:228 +0xea5
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:177 +0x56c
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x547
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.TestApfWatchPanic.func3()
      /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:553 +0xd0
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /usr/local/go/src/net/http/server.go:3071 +0x58
==================
    priority-and-fairness_test.go:180: Wanted 0 requests executing, got 1
    priority-and-fairness_test.go:595: expected panic, got <nil>
    testing.go:1312: race detected during execution of test
    --- FAIL: TestApfWatchHandlePanic/post-execute_panic (5.05s)
}
```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/110132/pull-kubernetes-unit/1527377549582143488

### What did you expect to happen?

test should not have data race

### How can we reproduce it (as minimally and precisely as possible)?

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/110132/pull-kubernetes-unit/1527377549582143488

This is the data race we saw in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125163/pull-kubernetes-unit/1814076004621094912 

```
k8s.io/apiserver/pkg/server: filters expand_less	1m19s
{Failed;Failed;  === RUN   TestApfWatchHandlePanic/post-execute_panic
==================
WARNING: DATA RACE
Read at 0x00000471d334 by goroutine 5478:
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:182 +0x324
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.WithRequestInfo.func4()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x17a
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.TestApfWatchHandlePanic.func3.1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:576 +0xc4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:3142 +0x2a1
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:2044 +0x13c4
  net/http.(*Server).Serve.gowrap3()
      /usr/local/go/src/net/http/server.go:3290 +0x4f

Previous write at 0x00000471d334 by goroutine 5477:
  sync/atomic.AddInt32()
      /usr/local/go/src/runtime/race_amd64.s:281 +0xb
  sync/atomic.AddInt32()
      <autogenerated>:1 +0x14
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle.func7()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:206 +0x1a7
  k8s.io/apiserver/pkg/server/filters.(*fakeWatchApfFilter).Handle()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:391 +0x127
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle.func8()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:264 +0x1e3

Goroutine 5478 (running) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3290 +0x8ec
  net/http/httptest.(*Server).goServe.func1()
      /usr/local/go/src/net/http/httptest/server.go:310 +0xbb

Goroutine 5477 (running) created at:
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:231 +0xfcc
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle-fm()
      <autogenerated>:1 +0x51
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:180 +0x309
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.WithRequestInfo.func4()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x17a
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.TestApfWatchHandlePanic.func3.1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:576 +0xc4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:3142 +0x2a1
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:2044 +0x13c4
  net/http.(*Server).Serve.gowrap3()
      /usr/local/go/src/net/http/server.go:3290 +0x4f
==================
    priority-and-fairness_test.go:183: Wanted 0 requests executing, got 1
    priority-and-fairness_test.go:573: expected panic, got <nil>
==================
WARNING: DATA RACE
Write at 0x00000471d334 by goroutine 5477:
  sync/atomic.AddInt32()
      /usr/local/go/src/runtime/race_amd64.s:281 +0xb
  sync/atomic.AddInt32()
      <autogenerated>:1 +0x14
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle.func7.deferwrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:207 +0x38
  runtime.deferreturn()
      /usr/local/go/src/runtime/panic.go:602 +0x5d
  k8s.io/apiserver/pkg/server/filters.(*fakeWatchApfFilter).Handle()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:391 +0x127
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle.func8()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:264 +0x1e3

Previous read at 0x00000471d334 by goroutine 5478:
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:183 +0x37a
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.WithRequestInfo.func4()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x17a
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.TestApfWatchHandlePanic.func3.1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:576 +0xc4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:3142 +0x2a1
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:2044 +0x13c4
  net/http.(*Server).Serve.gowrap3()
      /usr/local/go/src/net/http/server.go:3290 +0x4f

Goroutine 5477 (running) created at:
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:231 +0xfcc
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle-fm()
      <autogenerated>:1 +0x51
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:180 +0x309
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.WithRequestInfo.func4()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x17a
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  k8s.io/apiserver/pkg/server/filters.TestApfWatchHandlePanic.func3.1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:576 +0xc4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2171 +0x47
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:3142 +0x2a1
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:2044 +0x13c4
  net/http.(*Server).Serve.gowrap3()
      /usr/local/go/src/net/http/server.go:3290 +0x4f

Goroutine 5478 (finished) created at:
  net/http.(*Server).Serve()
      /usr/local/go/src/net/http/server.go:3290 +0x8ec
  net/http/httptest.(*Server).goServe.func1()
      /usr/local/go/src/net/http/httptest/server.go:310 +0xbb
==================
    testing.go:1398: race detected during execution of test
    --- FAIL: TestApfWatchHandlePanic/post-execute_panic (5.00s)
;=== RUN   TestApfWatchHandlePanic
--- FAIL: TestApfWatchHandlePanic (5.00s)
;}
```

the stack trace that corresponds to the source in k/k: https://github.com/kubernetes/kubernetes/blob/dbc2b0a5c7acc349ea71a14e49913661eaf708d2/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go#L182

--- a/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go
@@ -26,6 +26,7 @@ import (
 	"reflect"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -128,20 +129,20 @@ func newApfServerWithSingleRequest(t *testing.T, decision mockDecision) *httptes
 			t.Errorf("execute should not be invoked")
 		}
 		// atomicReadOnlyExecuting can be either 0 or 1 as we test one request at a time.
-		if decision != decisionSkipFilter && atomicReadOnlyExecuting != 1 {
-			t.Errorf("Wanted %d requests executing, got %d", 1, atomicReadOnlyExecuting)
+		if want, got := int32(1), atomic.LoadInt32(&atomicReadOnlyExecuting); decision != decisionSkipFilter && want != got {
+			t.Errorf("Wanted %d requests executing, got %d", want, got)
 		}
 	}
 	postExecuteFunc := func() {}
 	// atomicReadOnlyWaiting can be either 0 or 1 as we test one request at a time.
 	postEnqueueFunc := func() {
-		if atomicReadOnlyWaiting != 1 {
-			t.Errorf("Wanted %d requests in queue, got %d", 1, atomicReadOnlyWaiting)
+		if want, got := int32(1), atomic.LoadInt32(&atomicReadOnlyWaiting); want != got {
+			t.Errorf("Wanted %d requests in queue, got %d", want, got)
 		}
 	}
 	postDequeueFunc := func() {
-		if atomicReadOnlyWaiting != 0 {
-			t.Errorf("Wanted %d requests in queue, got %d", 0, atomicReadOnlyWaiting)
+		if want, got := int32(0), atomic.LoadInt32(&atomicReadOnlyWaiting); want != got {
+			t.Errorf("Wanted %d requests in queue, got %d", want, got)
 		}
 	}
 	return newApfServerWithHooks(t, decision, onExecuteFunc, postExecuteFunc, postEnqueueFunc, postDequeueFunc)
@@ -179,8 +180,8 @@ func newApfHandlerWithFilter(t *testing.T, flowControlFilter utilflowcontrol.Int
 		}))
 		apfHandler.ServeHTTP(w, r)
 		postExecute()
-		if atomicReadOnlyExecuting != 0 {
-			t.Errorf("Wanted %d requests executing, got %d", 0, atomicReadOnlyExecuting)
+		if want, got := int32(0), atomic.LoadInt32(&atomicReadOnlyExecuting); want != got {
+			t.Errorf("Wanted %d requests executing, got %d", want, got)
 		}
 	}), requestInfoFactory)
 
@@ -270,8 +271,8 @@ func TestApfExecuteMultipleRequests(t *testing.T) {
 	onExecuteFunc := func() {
 		preStartExecute.Done()
 		preStartExecute.Wait()
-		if int(atomicReadOnlyExecuting) != concurrentRequests {
-			t.Errorf("Wanted %d requests executing, got %d", concurrentRequests, atomicReadOnlyExecuting)
+		if want, got := int32(concurrentRequests), atomic.LoadInt32(&atomicReadOnlyExecuting); want != got {
+			t.Errorf("Wanted %d requests executing, got %d", want, got)
 		}
 		postStartExecute.Done()
 		postStartExecute.Wait()
@@ -280,8 +281,8 @@ func TestApfExecuteMultipleRequests(t *testing.T) {
 	postEnqueueFunc := func() {
 		preEnqueue.Done()
 		preEnqueue.Wait()
-		if int(atomicReadOnlyWaiting) != concurrentRequests {
-			t.Errorf("Wanted %d requests in queue, got %d", 1, atomicReadOnlyWaiting)
+		if want, got := int32(concurrentRequests), atomic.LoadInt32(&atomicReadOnlyWaiting); want != got {
+			t.Errorf("Wanted %d requests in queue, got %d", want, got)
 
 		}
 		postEnqueue.Done()
@@ -291,8 +292,8 @@ func TestApfExecuteMultipleRequests(t *testing.T) {
 	postDequeueFunc := func() {
 		preDequeue.Done()
 		preDequeue.Wait()
-		if atomicReadOnlyWaiting != 0 {
-			t.Errorf("Wanted %d requests in queue, got %d", 0, atomicReadOnlyWaiting)
+		if want, got := int32(0), atomic.LoadInt32(&atomicReadOnlyWaiting); want != got {
+			t.Errorf("Wanted %d requests in queue, got %d", want, got)
 		}
 		postDequeue.Done()
 		postDequeue.Wait()

--- a/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go
@@ -18,6 +18,7 @@ package filters
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"net/http"
 	"net/http/httptest"
@@ -178,11 +179,19 @@ func newApfHandlerWithFilter(t *testing.T, flowControlFilter utilflowcontrol.Int
 		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
 			Groups: []string{user.AllUnauthenticated},
 		}))
-		apfHandler.ServeHTTP(w, r)
-		postExecute()
-		if want, got := int32(0), atomic.LoadInt32(&atomicReadOnlyExecuting); want != got {
-			t.Errorf("Wanted %d requests executing, got %d", want, got)
-		}
+		func() {
+			// the defer ensures that the following assertion is
+			// executed, even if the APF handler panics
+			// TODO: all test(s) using this filter must run serially to each other
+			defer func() {
+				t.Logf("the APF handler has finished, checking atomicReadOnlyExecuting")
+				if want, got := int32(0), atomic.LoadInt32(&atomicReadOnlyExecuting); want != got {
+					t.Errorf("Wanted %d requests executing, got %d", want, got)
+				}
+			}()
+			apfHandler.ServeHTTP(w, r)
+			postExecute()
+		}()
 	}), requestInfoFactory)
 
 	return handler
@@ -346,19 +355,21 @@ func TestApfCancelWaitRequest(t *testing.T) {
 }
 
 type fakeWatchApfFilter struct {
+	t        *testing.T
 	lock     sync.Mutex
 	inflight int
 	capacity int
 
-	postExecutePanic bool
-	preExecutePanic  bool
+	postExecutePanic error
+	preExecutePanic  error
 
 	utilflowcontrol.WatchTracker
 	utilflowcontrol.MaxSeatsTracker
 }
 
-func newFakeWatchApfFilter(capacity int) *fakeWatchApfFilter {
+func newFakeWatchApfFilter(t *testing.T, capacity int) *fakeWatchApfFilter {
 	return &fakeWatchApfFilter{
+		t:               t,
 		capacity:        capacity,
 		WatchTracker:    utilflowcontrol.NewWatchTracker(),
 		MaxSeatsTracker: utilflowcontrol.NewMaxSeatsTracker(),
@@ -386,17 +397,23 @@ func (f *fakeWatchApfFilter) Handle(ctx context.Context,
 		return
 	}
 
-	if f.preExecutePanic {
-		panic("pre-exec-panic")
-	}
-	execFn()
-	if f.postExecutePanic {
-		panic("post-exec-panic")
-	}
+	func() {
+		defer func() {
+			f.lock.Lock()
+			defer f.lock.Unlock()
+			f.inflight--
+		}()
 
-	f.lock.Lock()
-	defer f.lock.Unlock()
-	f.inflight--
+		if f.preExecutePanic != nil {
+			f.t.Logf("going to panic (pre-exec) as expected with error: %v, fakeWatchApfFilter: %#v", f.preExecutePanic, f)
+			panic(f.preExecutePanic)
+		}
+		execFn()
+		if f.postExecutePanic != nil {
+			f.t.Logf("going to panic (post-exec) as expected with error: %v, fakeWatchApfFilter: %#v", f.postExecutePanic, f)
+			panic(f.postExecutePanic)
+		}
+	}()
 }
 
 func (f *fakeWatchApfFilter) Run(stopCh <-chan struct{}) error {
@@ -448,7 +465,7 @@ func TestApfExecuteWatchRequestsWithInitializationSignal(t *testing.T) {
 	allRunning := sync.WaitGroup{}
 	allRunning.Add(2 * concurrentRequests)
 
-	fakeFilter := newFakeWatchApfFilter(concurrentRequests)
+	fakeFilter := newFakeWatchApfFilter(t, concurrentRequests)
 
 	onExecuteFunc := func() {
 		firstRunning.Done()
@@ -494,7 +511,7 @@ func TestApfExecuteWatchRequestsWithInitializationSignal(t *testing.T) {
 }
 
 func TestApfRejectWatchRequestsWithInitializationSignal(t *testing.T) {
-	fakeFilter := newFakeWatchApfFilter(0)
+	fakeFilter := newFakeWatchApfFilter(t, 0)
 
 	onExecuteFunc := func() {
 		t.Errorf("Request unexepectedly executing")
@@ -513,7 +530,7 @@ func TestApfWatchPanic(t *testing.T) {
 	epmetrics.Register()
 	fcmetrics.Register()
 
-	fakeFilter := newFakeWatchApfFilter(1)
+	fakeFilter := newFakeWatchApfFilter(t, 1)
 
 	onExecuteFunc := func() {
 		panic("test panic")
@@ -540,11 +557,11 @@ func TestApfWatchPanic(t *testing.T) {
 func TestApfWatchHandlePanic(t *testing.T) {
 	epmetrics.Register()
 	fcmetrics.Register()
-	preExecutePanicingFilter := newFakeWatchApfFilter(1)
-	preExecutePanicingFilter.preExecutePanic = true
+	preExecutePanicingFilter := newFakeWatchApfFilter(t, 1)
+	preExecutePanicingFilter.preExecutePanic = http.ErrAbortHandler
 
-	postExecutePanicingFilter := newFakeWatchApfFilter(1)
-	postExecutePanicingFilter.postExecutePanic = true
+	postExecutePanicingFilter := newFakeWatchApfFilter(t, 1)
+	postExecutePanicingFilter.postExecutePanic = http.ErrAbortHandler
 
 	testCases := []struct {
 		name   string
@@ -560,18 +577,31 @@ func TestApfWatchHandlePanic(t *testing.T) {
 		},
 	}
 
-	onExecuteFunc := func() {
-		time.Sleep(5 * time.Second)
-	}
-	postExecuteFunc := func() {}
-
 	for _, test := range testCases {
 		t.Run(test.name, func(t *testing.T) {
+			onExecuteFunc := func() {
+				time.Sleep(5 * time.Second)
+
+				// this function should not be executed if
+				// pre-execute panic is set
+				if test.filter.preExecutePanic != nil {
+					t.Errorf("did not expect the execute function to be executed")
+				}
+				t.Logf("on-execute function invoked")
+			}
+
+			// we either panic before the execute function, or after,
+			// so the following function should never be executed.
+			postExecuteFunc := func() {
+				t.Errorf("did not expect the post-execute function to be invoked")
+			}
+
 			apfHandler := newApfHandlerWithFilter(t, test.filter, time.Minute/4, onExecuteFunc, postExecuteFunc)
 			handler := func(w http.ResponseWriter, r *http.Request) {
 				defer func() {
-					if err := recover(); err == nil {
-						t.Errorf("expected panic, got %v", err)
+					recovered := recover()
+					if err, ok := recovered.(error); !ok || !errors.Is(err, http.ErrAbortHandler) {
+						t.Errorf("expected panic with error: %v, but got: %v", http.ErrAbortHandler, err)
 					}
 				}()
 				apfHandler.ServeHTTP(w, r)

